src/jit/optimizer.cpp

```diff
@@ -6805,24 +6805,33 @@ void Compiler::optHoistLoopExprsForBlock(BasicBlock* block, unsigned loopNum, Lo
             UseExecutionOrder = true,
         };
 
-        HoistVisitor(Compiler* compiler, bool firstBlock, unsigned loopNum, LoopHoistContext* hoistContext)
+        HoistVisitor(Compiler* compiler, unsigned loopNum, LoopHoistContext* hoistContext)
             : GenTreeVisitor(compiler)
             , m_valueStack(compiler->getAllocator(CMK_LoopHoist))
-            , m_beforeSideEffect(firstBlock)
+            , m_beforeSideEffect(true)
             , m_loopNum(loopNum)
             , m_hoistContext(hoistContext)
         {
         }
 
-        void Hoist(GenTree* tree)
+        void HoistBlock(BasicBlock* block)
         {
-            WalkTree(&tree, nullptr);
-
-            if (m_valueStack.TopRef().m_hoistable)
+            for (GenTreeStmt* stmt = block->FirstNonPhiDef(); stmt != nullptr; stmt = stmt->gtNextStmt)
             {
-                assert(m_valueStack.TopRef().Node() == tree);
-                m_compiler->optHoistCandidate(tree, m_loopNum, m_hoistContext);
+                WalkTree(&stmt->gtStmtExpr, nullptr);
+                assert(m_valueStack.TopRef().Node() == stmt->gtStmtExpr);
+
+                if (m_valueStack.TopRef().m_hoistable)
+                {
+                    m_compiler->optHoistCandidate(stmt->gtStmtExpr, m_loopNum, m_hoistContext);
+                }
+
+                m_valueStack.Reset();
             }
+
+            // We don't visit every block in the loop so once we visited the first block
```

Comment:
I do not understand this comment, could you please clarify? 
Does it mean that we haven't used each block yet? Or does it mean that we are skipping some blocks? Why is  the first block special?

Reply:
The list of blocks that `optHoistThisLoop` passes contains only blocks known to be definitely executed so basically there are skipped blocks. Those can have side effects that we do not know about. Maybe I should have just moved the block list building code here but for now I'll just expand this comment a bit, there's enough code movement for one PR.